File: Examples/Java/Containers/MongoDB-REST/src/main/java/com/api/mongodb/rest/service/mongodb/crud/MongoDB_CRUD_Client.java
package com.api.mongodb.rest.service.mongodb.crud;

import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.model.Filters;

import org.bson.BsonDocument;
import org.bson.conversions.Bson;
import org.bson.types.ObjectId;
import java.util.concurrent.CompletableFuture;

public class MongoDB_CRUD_Client implements IMongoDB_CRUD_Client {
    static private String _connectionString;
    static private MongoClient _mongoClient;

    static public void initialize(String connectionString) {
        _connectionString = connectionString;
    }

    static private MongoClient getMongoClient() { 
        if (null == _mongoClient)
            _mongoClient = MongoClients.create(_connectionString);
        return _mongoClient; 
    }

    private MongoCollection<BsonDocument> getCollection(String dbName, String collectionName) {
        MongoDatabase database = getMongoClient().getDatabase(dbName);
        return database.getCollection(collectionName, BsonDocument.class);
    }

    public CompletableFuture<Void> ping() {
        return CompletableFuture.runAsync(() -> {
            getMongoClient().getDatabase("admin").runCommand(new BsonDocument("ping", new BsonDocument()));
        });
    }

    public CompletableFuture<ObjectId> createAsync(String dbName, String collectionName, BsonDocument document) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            collection.insertOne(document);
            return document.getObjectId("_id").getValue();
        });
    }

    public CompletableFuture<BsonDocument> readAsync(String dbName, String collectionName, ObjectId documentId) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            return collection.find(filter).first();
        });
    }

    public CompletableFuture<BsonDocument> updateAsync(String dbName, String collectionName, ObjectId documentId, BsonDocument updateFields) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            BsonDocument update = new BsonDocument("$set", updateFields);
            return collection.findOneAndUpdate(filter, update);
        });
    }

    public CompletableFuture<Boolean> deleteAsync(String dbName, String collectionName, ObjectId documentId) {
        return CompletableFuture.supplyAsync(() -> {
            MongoCollection<BsonDocument> collection = getCollection(dbName, collectionName);
            Bson filter = Filters.eq("_id", documentId);
            return collection.deleteOne(filter).getDeletedCount() == 1;
        });
    }    
}

File: Examples/CSharp/Containers/MongoDB-REST/MongoDB.REST/MongoDB/CRUD/MongoDB_CRUD_Client.cs
namespace MongoDB.CRUD.Client;

using MongoDB.Bson;
using MongoDB.Driver;

public class MongoDB_CRUD_Client(IMongoClient mongoClient) : IMongoDB_CRUD_Client
{
    private IMongoClient MongoClient { get; } = mongoClient;

    public async Task Ping()
    {
        await MongoClient.GetDatabase("admin").RunCommandAsync<BsonDocument>(new BsonDocument("ping", 1));
    }

    private IMongoCollection<BsonDocument> GetCollection(string dbName, string collectionName)
    {
        var database = MongoClient.GetDatabase(dbName);
        return database.GetCollection<BsonDocument>(collectionName);
    }

    public async Task<ObjectId?> CreateAsync(string dbName, string collectionName, BsonDocument document, CancellationToken cancellationToken = default)
    {
        var collection = GetCollection(dbName, collectionName);
        await collection.InsertOneAsync(document, null, cancellationToken);
        return document["_id"].AsObjectId;
    }

    public async Task<BsonDocument?> ReadAsync(string dbName, string collectionName, ObjectId documentId, CancellationToken cancellationToken = default)
    {
        var collection = GetCollection(dbName, collectionName);
        var filter = Builders<BsonDocument>.Filter.Eq("_id", documentId);
        return (await collection.FindAsync(filter, null, cancellationToken)).FirstOrDefault(cancellationToken);        
    }

    public async Task<BsonDocument?> UpdateAsync(string dbName, string collectionName, ObjectId documentId, BsonDocument updateFields, CancellationToken cancellationToken = default)
    {
        var collection = GetCollection(dbName, collectionName);
        var filter = Builders<BsonDocument>.Filter.Eq("_id", documentId);
        var update = Builders<BsonDocument>.Update
                    .Combine(updateFields.Elements.Select(e => Builders<BsonDocument>.Update.Set(e.Name, e.Value)));
        return await collection.FindOneAndUpdateAsync(filter, update, new FindOneAndUpdateOptions<BsonDocument> { ReturnDocument = ReturnDocument.After }, cancellationToken);
    }

    public async Task<bool> DeleteAsync(string dbName, string collectionName, ObjectId documentId, CancellationToken cancellationToken = default)
    {
        var collection = GetCollection(dbName, collectionName);
        var filter = Builders<BsonDocument>.Filter.Eq("_id", documentId);
        var result = await collection.DeleteOneAsync(filter, cancellationToken);
        return result.DeletedCount == 1;
    }
}